internal/common/spiretrustsource.go

```diff
@@ -0,0 +1,165 @@
+package common
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/workload"
+)
+
+// SpireTrustSource holds all necessary information to connect to a spire instance and store its certificates
+type SpireTrustSource struct {
+	domainURLs         map[string]string
+	domainCertificates map[string][]*x509.Certificate
+	spireClients       []*workload.X509SVIDClient
+	certLocation       string
+}
+
+type certMap struct {
+	Certs map[string][]string `json:"certs"`
+}
+
+type watcher struct {
+	uri    string
+	source *SpireTrustSource
+}
+
+// TrustedCertificates fulfills the TrustSource interface
+func (s *SpireTrustSource) TrustedCertificates() map[string][]*x509.Certificate {
+	return s.domainCertificates
+}
+
+// NewSpireTrustSource creates a new trust source with connectivity to one or more spire instances.
+func NewSpireTrustSource(domainURLs map[string]string, certLocation string) (*SpireTrustSource, error) {
+	source := SpireTrustSource{
+		domainURLs:         domainURLs,
+		domainCertificates: make(map[string][]*x509.Certificate, 0),
+		certLocation:       certLocation,
+	}
+
+	if certLocation != "" {
+		if err := source.parseCertFile(); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := source.startWatchers(); err != nil {
+		return nil, err
+	}
+
+	return &source, nil
+}
+
+// Stop stops all spire clients and writes out certs
+func (s *SpireTrustSource) Stop() {
+	for _, client := range s.spireClients {
+		client.Stop()
+	}
+	s.writeCertFile()
+}
+
+func (s *SpireTrustSource) parseCertFile() error {
+	file, err := appFS.OpenFile(s.certLocation, os.O_RDWR|os.O_CREATE, 600)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileDat, err := ioutil.ReadAll(file)
+	if err != nil {
+		return fmt.Errorf("could not read cert file: %v", err)
+	}
+	fmt.Println(string(fileDat))
+	var certStruct certMap
+	err = json.Unmarshal(fileDat, &certStruct)
```

Comment:
you can compress this:
```suggestion
	if err = json.Unmarshal(fileDat, &certStruct); err != nil {
         	logrus.Warnf("Error unmarshaling cert file: %v\n", err) 
      }
```